lib/Conversion/AffineToPipeline/AffineToPipeline.cpp

```diff
@@ -57,20 +57,73 @@ struct AffineToPipeline : public AffineToPipelineBase<AffineToPipeline> {
   LogicalResult populateOperatorTypes(SmallVectorImpl<AffineForOp> &loopNest);
   LogicalResult solveSchedulingProblem(SmallVectorImpl<AffineForOp> &loopNest);
   LogicalResult createPipelinePipeline(SmallVectorImpl<AffineForOp> &loopNest);
+  LogicalResult constantCSE();
+  LogicalResult unrollAndForwardStores();
 
   CyclicSchedulingAnalysis *schedulingAnalysis;
 };
 
+static LogicalResult
+forwardStoreToLoad(AffineReadOpInterface loadOp,
+                   SmallVectorImpl<Operation *> &opsToErase) {
+
+  Operation *prevOp = loadOp->getPrevNode();
+  while (prevOp) {
+    if (prevOp->getParentRegion() != loadOp->getParentRegion())
+      return failure();
+
+    auto storeOp = dyn_cast<AffineWriteOpInterface>(prevOp);
+    if (!storeOp) {
+      prevOp = prevOp->getPrevNode();
+      continue;
+    }
+
+    if (storeOp.getMemRef() != loadOp.getMemRef())
+      return failure();
```

Comment:
I would recommend doing a check that says: "If something between these two operations tries to touch memref, fail". This is essentially what `hasNoInterveningEffects` does, but since you have a problem with the dependence analysis being expensive (although I have my doubts about the claim that it is slow. It is pretty fast AFAIU), you fail if anything even uses memref.

This would make it a conservative check instead of a hack. 

I would also add a comment saying this is a conservative check which works well in unrolling cases but in general, may not work well.

Reply:
Rather than hacking some subset of `hasNoInterveningEffects`, now I just call `hasNoInterveningEffects` as such:

```cpp
    if (!mlir::hasNoInterveningEffect<MemoryEffects::Write>(storeOp, loadOp)
        || !mlir::hasNoInterveningEffect<MemoryEffects::Read>(storeOp, loadOp))
      return failure();
```

Only issue is now this PR depends on my [upstream LLVM PR](https://reviews.llvm.org/D134374). Hence, @mikeurbach if you want to checkout you'll need to [apply this patch](https://reviews.llvm.org/file/data/zvi3mhhyojhn2lmseijm/PHID-FILE-6wyulgp7aywzdklzzsaz/D134374.diff) to `llvm/`.